src/platforms/stm32/src/lib/gpio_driver.c

```diff
@@ -240,117 +262,87 @@ static term setup_gpio_pin(Context *ctx, term gpio_pin_tuple, term mode_term)
 
     if (UNLIKELY(!term_is_tuple(gpio_pin_tuple))) {
         AVM_LOGE(TAG, "Invalid GPIO Pin tuple, expect {Bank, Pin}.");
-        return create_pair(ctx, ERROR_ATOM, BADARG_ATOM);
+        RAISE_ERROR(BADARG_ATOM);
```

Comment:
So I think it is worth having an important comment/caveat in the heading of the setup_gpio_pin function to the effect that the return value of this function MUST be immediately returned from any Nif function, the reason being that the RAISE_ERROR macro has side-effects on the Context structure that could have unintended consequences if not handled properly.  It _is_ handled properly in this module, but for future maintainers, it is worth a note (IMO) to the effect of the above, so that this function is not mis-used.

Reply:
Maybe we should also document this macro in the API docs? In a separate PR, of course. It can be kind of difficult to know when (and how) some of the macros should be used, since most are undocumented.

Reply:
Maybe it would make more sense to put this warning at the top of the nif handling section of the code (since most can potentially RAISE_ERROR) and that is the place where we need to make sure the function results are returned directly.

Reply:
Is there also a side effect for ports? If so, the use of RAISE_ERROR is not the best choice here.

Reply:
I think we don't (or shouldn't use this in ports, since we don't have access to the context stack from there.  But I agree that a comment is worth while in the nif header.  Also, I think there may be multiple definitions of the macro (one in opcodesswitch, for example)

Reply:
OK. Then these changes are not good at all, because for simplicity, both the nif and port implementation call the same functions for setup, read, write, etc.

Reply:
I will revert this to `{error, Reason}`, and do some extra processing in the nif implementation to turn that into an appropriate `RAISE_ERROR(Error)` in the nif handler.